### PR TITLE
feat: hook up product tutorials sidecar

### PR DESCRIPTION
**Pull request by zchsh**
## ✅ "Ready for Review" checklist

- [x] The Vercel preview link has been added to this PR's description
- [x] The Asana task has been added to this PR's description
- [x] This PR's link has been added to the "📐 Pull Request" field on the Asana task
- [x] This Vercel preview link has been added to the "🔍 Preview" field on the Asana task
- [x] The description template has been filled in below

---

## 🔗 Relevant links

- [Preview link][preview] 🔎
- [Asana task][task] 🎟️

## 🗒️ What

Implements `Sidecar`-compatible `heading` elements in `ProductTutorialsView`. 

## 🤷 Why

Because we want the sidecar on `ProductTutorialsView` to function as expected.

## 🛠️ How

- Immediately after getting pageData from the learn client, add `headingSlug` values to it
- Use these `headingSlug` values for both the `<heading id={slug} />` element rendering, and the `Sidecar`

## 📸 Design Screenshots

Not applicable - sidecar is already styled, this PR implements functionality specifically for the `ProductTutorialsView`.

## 🧪 Testing

- [ ] Visit [`/waypoint/tutorials`][preview] and ensure each sidecar jump-links works
- [ ] Visit [`/vault/tutorials`][preview-vault] and ensure each sidecar jump-links works

>  ℹ️ Note: sidecar active states don't seem to be fully functional. Clicking on a sidecar link should jump the page, but the positioning and highlighting of active items may not be correct. I think the latter issues are outside the scope of this PR. We have a task to dive deeper into scroll issues (looks like it may be `scroll-margin-top` related, may be able to handle in follow-up UI implementation PRs): 🎟️ [bug: sidecar anchor links highlight in odd ways in some cases](https://app.asana.com/0/1202097197789424/1202125773561669/f)

## 💭 Anything else?

Not at the moment!

[preview]: https://dev-portal-git-zsproduct-tutorials-sidecar-hashicorp.vercel.app/waypoint/tutorials
[preview-vault]: https://dev-portal-git-zsproduct-tutorials-sidecar-hashicorp.vercel.app/waypoint/tutorials
[task]: https://app.asana.com/0/1202097197789424/1202182325935218/f

### CODE REVIEW

File: src/views/product-tutorials-view/components/product-view-content/types.ts [hunk 1]
```diff
@@ -10,8 +10,7 @@ import {
   TutorialsStackProps,
 } from './components'
 
-interface FeaturedStackBlock
-  extends Pick<FeaturedStackProps, 'heading' | 'subheading'> {
+interface FeaturedStackBlock extends Omit<FeaturedStackProps, 'children'> {
```
Comment (zchsh) [inline src/views/product-tutorials-view/components/product-view-content/types.ts:13]: I think `Omit<>` might make more sense here, especially with `headingSlug` in place. I think I'd initially done `Pick<>` since I'd just come across it in my lil TypeScript journey 😅

File: src/views/product-tutorials-view/components/sitemap/index.tsx [hunk 2]
```diff
@@ -9,7 +9,7 @@ import s from './sitemap.module.css'
 export function ProductTutorialsSitemap({
   collections,
 }: {
-  collections: ProductPageData['pageData']['allCollections']
+  collections: ProductPageData['allCollections']
```
Comment (zchsh) [inline src/views/product-tutorials-view/components/sitemap/index.tsx:12]: Have moved `allCollections` up a level - I think it's helpful if we keep `pageData` as authored, and enhance it (in the front-end for now) with stuff like `inlineCollections`, `inlineTutorials`, and `allCollections`.

File: src/views/product-tutorials-view/helpers/heading-helpers.ts [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+   */
+  return [...overviewHeading, ...blockHeadings, ...sitemapHeading]
+}
+
+/**
+ * Given a productName,
+ * return a TableOfContentsHeading object
+ * representing a top-level heading for a product tutorials view
+ */
+function getOverviewHeading(productName: string): TableOfContentsHeading {
```
Comment (zchsh) [inline src/views/product-tutorials-view/helpers/heading-helpers.ts:61]: I've split out `getOverviewHeading` and `getSitemapHeading` to make it more clear that they're used both server-side (to generate the array of `headings` for the sidecar), as well as client-side (to make sure the headings being linked to get rendered).

File: src/components/collection-card/components/company-logo/index.tsx [hunk 4]
```diff
@@ -3,22 +3,28 @@ import classNames from 'classnames'
 import { CompanyLogoOption, CompanyLogoProps } from './types'
 import s from './company-logo.module.css'
 
+const LOGO_TO_SOURCE_MAP: Record<CompanyLogoOption, string> = {
```
Comment (zchsh) [inline src/components/collection-card/components/company-logo/index.tsx:6]: I've moved `CompanyLogoOption` into `learn-client/types`. This felt more natural when constructing the `ProductPage` block types, and it felt intuitive to think of our component as filling out its logo options based on possibilities defined by `learn-client`, rather than the other way around.

On a related note, I've set up `LOGO_TO_SOURCE_MAP` for a similar reason - the intent is better type safety, as we'll now get a warning if `CompanyLogoOption` changes but we haven't account for it with an explicit image path.

File: src/lib/learn-client/types.ts [hunk 5]
```diff
@@ -118,10 +118,65 @@ export type AllCollectionsProductOptions = {
...
+  | ProductPageBlockBrandedCallout
+  | ProductPageBlockCardList
+  | ProductPageBlockTutorialsStack
+  | ProductPageBlockCollectionsStack
+
 export interface ProductPage {
   slug: string
   pageData: {
-    blocks: any[] // @TODO update when we add in blocks
+    blocks: ProductPageBlock[]
```
Comment (zchsh) [inline src/lib/learn-client/types.ts:179]: Have done my best to resolve this TODO by building out the types above 👍

File: src/lib/learn-client/types.ts [hunk 6]
```diff
@@ -188,6 +243,16 @@ export enum HandsOnLabProviderOption {
   instruqt = 'instruqt',
 }
 
+export enum CompanyLogoOption {
```
Comment (zchsh) [inline src/lib/learn-client/types.ts:246]: Even though this is currently specific to `ProductPage` data, the miscellaneous enums section felt like a good home.

File: src/pages/vault/tutorials/index.tsx [hunk 7]
```diff
@@ -1,15 +1,15 @@
 import productData from 'data/vault.json'
 import { LearnProductData } from 'types/products'
 import {
-  getProductTutorialsPageProps,
-  ProductTutorialsPageProps,
+  getProductTutorialsViewProps,
+  ProductTutorialsViewProps,
```
Comment (zchsh) [inline src/pages/vault/tutorials/index.tsx:5]: Have renamed this in an effort to more cleanly differentiate between `ProductPageBlock`, the "block" type we receive from the learn client, and `ProductViewBlock`, the "block" type we pass to components in the front end. This is definitely a bit of a half-measure (I feel like the names are maybe just _slightly_ more differentiated), so very open to input here 🙇

File: src/views/product-tutorials-view/components/product-view-content/components/branded-callout/index.tsx [hunk 8]
```diff
@@ -3,26 +3,30 @@ import s from './branded-callout.module.css'
 
 function BrandedCallout({
   heading,
+  headingSlug,
```
Comment (zchsh) [inline src/views/product-tutorials-view/components/product-view-content/components/branded-callout/index.tsx:6]: All top-level `product-view-content` blocks now support a `headingSlug` prop, which keeps them in sync with the `#heading-slug` links we're rendering in the `Sidecar`.

File: src/views/product-tutorials-view/helpers/generate-headings.ts [hunk 9]
```diff
@@ -1,18 +0,0 @@
-import slugify from 'slugify'
-
-//@TODO type blocks
-export function generateHeadings(blocks: any) {
```
Comment (zchsh) [inline src/views/product-tutorials-view/helpers/generate-headings.ts:4]: Started from this function, and built out `heading-helpers` based on it.

File: src/views/product-tutorials-view/index.tsx [hunk 10]
```diff
@@ -31,20 +41,24 @@ function ProductTutorialsView({
...
       />
-      {showProductSitemap ? (
-        <>
-          <h2 id={layoutProps.headings[layoutProps.headings.length - 1].slug}>
-            All tutorials
-          </h2>
-          <ProductTutorialsSitemap collections={allCollections} />
-        </>
-      ) : null}
+      {showProductSitemap
```
Comment (zchsh) [inline src/views/product-tutorials-view/index.tsx:50]: The inline anonymous function here is kinda gritty. Open to moving this; my hot take was we could live with the awkwardness for now and move this into the `sitemap` component itself in the subsequent PR where we implement `sitemap` to spec.

File: src/views/product-tutorials-view/server.ts [hunk 11]
```diff
@@ -33,10 +37,10 @@ type ProductTutorialsLayout = Pick<
 
 export interface ProductPageData {
   pageData: {
-    blocks: any // @TODO type
-    showProductSitemap: boolean
-    allCollections: ClientCollection[]
+    blocks: ProductViewBlock[]
+    showProductSitemap?: boolean
   }
+  allCollections: ClientCollection[]
```
Comment (zchsh) [inline src/views/product-tutorials-view/server.ts:43]: As mentioned in earlier comment, no tangible change to `allCollections`; moved it alongside `pageData` for clarity in authorable vs "augmenting" data.

File: src/views/product-tutorials-view/server.ts [hunk 12]
```diff
@@ -51,12 +55,26 @@ export interface ProductPageData {
...
+   */
+  const {
+    pageData: rawPageData,
+    inlineCollections,
+    inlineTutorials,
+  } = await getProductPageContent(productSlug)
+  /**
+   * Add headings to raw page data, for use with the page's sidecar
+   */
+  const pageData = addHeadingSlugsToBlocks(rawPageData)
```
Comment (zchsh) [inline src/views/product-tutorials-view/server.ts:73]: 🌟 This probably best captures the broad intent of the PR, being:
- Immediately after getting pageData from the learn client, add `headingSlug` values to it
- Use these `headingSlug` values for both the `<heading id={slug} />` element rendering, and the `Sidecar`

File: src/components/collection-card/index.tsx [hunk 13]
```diff
@@ -1,13 +1,12 @@
...
-import classNames from 'classnames'
 import CardLink from 'components/card-link'
 import {
   CardEyebrow,
   CardHeading,
   CardBody,
   CardBadges,
 } from 'components/tutorial-collection-cards'
-import { CollectionCardProps } from './types'
+import { CollectionCardProps, CompanyLogoOption } from './types'
```
Comment (zchsh) [inline src/components/collection-card/index.tsx:9]: Threading `CompanyLogoOption` thru to swingset.

File: src/components/collection-card/components/company-logo/types.ts [hunk 14]
```diff
@@ -1,12 +1,6 @@
-export enum CompanyLogoOption {
-  docker = 'docker',
-  github = 'github',
-  'microsoft-azure' = 'microsoft-azure',
-  oci = 'oci',
-  'google-cloud' = 'google-cloud',
-  'terraform-cloud' = 'terraform-cloud',
-  aws = 'aws',
-}
+import { CompanyLogoOption } from 'lib/learn-client/types'
```
Comment (zchsh) [inline src/components/collection-card/components/company-logo/types.ts:1]: I've moved `CompanyLogoOption` into `learn-client/types`. This felt more natural when constructing the `ProductPage` block types, and it felt intuitive to think of our component as filling out its logo options based on possibilities defined by `learn-client`, rather than the other way around.

Definitely happy to adjust here though, I can see it both ways 🤔

File: src/components/collection-card/docs.mdx [hunk 15]
```diff
@@ -28,15 +28,7 @@ With logos as heading:
...
-  {[
-    'docker',
-    'github',
-    'microsoft-azure',
-    'oci',
-    'google-cloud',
-    'terraform-cloud',
-    'aws',
-  ].map((logo, index) => {
+  {Object.keys(CompanyLogoOption).map((logo, index) => {
```
Comment (zchsh) [inline src/components/collection-card/docs.mdx:31]: Tangential to the PR but including cause I messed with `CompanyLogoOption` in the process of doing TypeScript pre-work for `heading` stuff: I've threaded through `CompanyLogoOption` as a named export of `collection-card`, so that we can iterate over the enum in Swingset. This way, if we ever miss an option, I _think_ we should see a failed build rather than pushing something broken to production.

File: src/components/tutorial-collection-cards/components/card-heading/company-logo/index.tsx [hunk 16]
```diff
@@ -1,28 +0,0 @@
-import Image from 'next/image'
```
Comment (zchsh) [inline src/components/tutorial-collection-cards/components/card-heading/company-logo/index.tsx:1]: Pretty tangential to the PR, but during type work I moved `CompanyLogoOption` around, and that brought to my attention these unused sub-components of `card-heading`.

Happy to drop these from this PR & handle in a follow-up if we wanna do that instead 🙇

File: src/components/tutorial-collection-cards/components/card-heading/types.ts [hunk 17]
```diff
@@ -1,5 +1,4 @@
 import type { HeadingProps } from 'components/heading'
-import { CompanyLogoOption } from './company-logo/types'
```
Comment (zchsh) [inline src/components/tutorial-collection-cards/components/card-heading/types.ts:2]: Not actually used, all `CompanyLogo` stuff is specific to `collection-card` and was moved there in a previous PR. I just forgot to delete it.

File: src/views/product-tutorials-view/components/product-view-content/components/logo-card-list/types.ts [hunk 18]
```diff
@@ -1,6 +1,5 @@
 import { Collection as ClientCollection } from 'lib/learn-client/types'
-
-import { CompanyLogoOption } from 'components/collection-card/components/company-logo/types'
+import { CompanyLogoOption } from 'lib/learn-client/types'
```
Comment (zchsh) [inline src/views/product-tutorials-view/components/product-view-content/components/logo-card-list/types.ts:2]: importing from the original (now moved) type location rather than an alias

**Comment by zchsh** [comment]
@kendallstrautman This PR gets the `Sidecar` working on `ProductTutorialsView`.

Context: I was scoping out the `ProductTutorialsView` work this AM and realized this is probably one of the hairier pieces (the rest should mostly be chroming some components, I think). Apologies for the relatively large PR; happy to add further context wherever needed 🙇

File: src/components/collection-card/components/company-logo/types.ts (hunk 14 above)
Comment (kendallstrautman) [reply src/components/collection-card/components/company-logo/types.ts:1]: I agree, it makes sense there 👌

File: src/lib/learn-client/types.ts [hunk 19]
```diff
@@ -118,10 +118,65 @@ export type AllCollectionsProductOptions = {
   sidebarSort?: boolean
 }
 
+/**
+ * PRODUCT LANDING PAGES -------------------------------
+ */
+
+export interface ProductPageBlockBrandedCallout {
```
Comment (kendallstrautman) [inline src/lib/learn-client/types.ts:125]: 💭 I wonder if the `ProductPage` prefix needs to be even _more_ explicit, since its tutorials specific. I'm not sure its necessary but I _could_ see room for confusion on what product page is referring to.

File: src/pages/vault/tutorials/index.tsx (hunk 7 above)
Comment (kendallstrautman) [reply src/pages/vault/tutorials/index.tsx:5]: Works for me!

File: src/views/product-tutorials-view/components/product-view-content/types.ts (hunk 1 above)
Comment (kendallstrautman) [reply src/views/product-tutorials-view/components/product-view-content/types.ts:13]: Love it 💯 Agreed it's definitely cleaner when you only have one thing to remove.

Comment (kendallstrautman) [reply src/views/product-tutorials-view/components/product-view-content/types.ts:13]: Sometimes it depends though and its easier to read `Pick`, since you can scan what is being extended. Just something to think about.

File: src/views/product-tutorials-view/index.tsx (hunk 10 above)
Comment (kendallstrautman) [reply src/views/product-tutorials-view/index.tsx:50]: Agreed, creating its own component when you go to style makes sense.

**Review by kendallstrautman — APPROVED** [review]
👍

File: src/views/product-tutorials-view/components/product-view-content/types.ts (hunk 1 above)
Comment (zchsh) [reply src/views/product-tutorials-view/components/product-view-content/types.ts:13]: ah yeah great point - now I'm on the fence again 😝 might re-think this again as I work through the implementation of individual components 😅

File: src/lib/learn-client/types.ts (hunk 19 above)
Comment (zchsh) [reply src/lib/learn-client/types.ts:125]: Ooh great point 🤔

For now I took another look at this, and realized I really like the pattern of `import { Thing as ClientThing }` I've seen in other areas of the app where we import `learn-client/types` stuff into front-end page & component files. So I'm doing eg `import { ProductPageBlock as LearnClientProductPageBlock } from 'lib/learn-client/types'`. A bit wordy but hopefully feels more consistent with other imports ✨ 

Definitely open to iterating on the naming here further - will keep it in mind as I work through individual components; I think if I end up extending & otherwise re-using the `ProductPageBlock<BlockType>` types, it'll probably come up again ♻️
